python/lsst/ip/isr/isrMock.py

```diff
@@ -0,0 +1,915 @@
+            u = x + amp.getBBox().getBeginX()
+            v = y + amp.getBBox().getBeginY()
+        else:
+            u = x + amp.getRawDataBBox().getBeginX()
+            v = y + amp.getRawDataBBox().getBeginY()
+
+        return (v, u)
+
+    # Simple data values.
+    def amplifierAddNoise(self, ampData, mean, sigma):
+        r"""Add Gaussian noise to an amplifier's image data.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        mean : float
+            Mean value of the Gaussian noise.
+        sigma : float
+            Sigma of the Gaussian noise.
+        """
+        ampArr = ampData.getArray()
+        ampArr[:] = ampArr[:] + self.rng.normal(mean, sigma,
+                                                size=ampData.getDimensions()).transpose()
+
+    def amplifierAddYGradient(self, ampData, start, end):
+        r"""Add a y-axis linear gradient to an amplifier's image data.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        start : float
+            Start value of the gradient (at y=0).
+        end : float
+            End value of the gradient (at y=ymax).
+        """
+        nPixY = ampData.getDimensions().getY()
+        ampArr = ampData.getArray()
+        ampArr[:] = ampArr[:] + (np.interp(range(nPixY), (0, nPixY - 1), (start, end)).reshape(nPixY, 1) +
+                                 np.zeros(ampData.getDimensions()).transpose())
+
+    def amplifierAddSource(self, ampData, scale):
+        r"""Add a single Gaussian source to an amplifier.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        scale : float
+            Peak flux of the source to add.
+        """
+        for x in range(0, ampData.getDimensions().getX()):
+            for y in range(0, ampData.getDimensions().getY()):
+                ampData.getArray()[y][x] = (ampData.getArray()[y][x] +
+                                            scale * np.exp(-0.5 * ((x - 50)**2 + (y - 25)**2)**2 / 3.0**2))
+
+    def amplifierAddCT(self, ampDataSource, ampDataTarget, scale):
+        r"""Add a scaled copy of an amplifier to another, simulating crosstalk.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        scale : float
+            Peak flux of the source to add.
+
+        Notes
+        -----
+        This simulates simple crosstalk between amplifiers.
+        """
+        for x in range(0, ampDataSource.getDimensions().getX()):
```

Comment:
Surely this is doable with a one-liner, _i.e._ not looping over pixels?

Reply:
Something like
`ampDataTarget.array[:] +=  scale * ampDataSource.array[:]`
replacing the whole loop? Or am I not thinking about something here?